traffic_portal/app/src/common/service/utils/TopologyUtils.js

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+var TopologyUtils = function() {
+
+	let normalizedTopology;
+
+	let flattenTopology = function(topologyTree, fromScratch) {
+		if (fromScratch) normalizedTopology.nodes = [];
+		topologyTree.forEach(function(node) {
+			if (node.cachegroup) {
+				normalizedTopology.nodes.push({
+					cachegroup: node.cachegroup,
+					parent: node.parent,
+					secParent: node.secParent,
+					parents: []
+				});
+			}
+			if (node.children && node.children.length > 0) {
+				flattenTopology(node.children, false);
+			}
+		});
+	};
+
+	let addNodeIndexes = function() {
+		normalizedTopology.nodes.forEach(function(currentNode) {
+			let parentNodeIndex = _.findIndex(normalizedTopology.nodes, function(node) { return currentNode.parent === node.cachegroup });
+			let secParentNodeIndex = _.findIndex(normalizedTopology.nodes, function(node) { return currentNode.secParent === node.cachegroup });
+			if (parentNodeIndex > -1) {
+				currentNode.parents.push(parentNodeIndex);
+				if (secParentNodeIndex > -1) {
+					currentNode.parents.push(secParentNodeIndex);
+				}
+			}
+		});
+	};
+
+	this.getNormalizedTopology = function(name, description, topologyTree) {
+		// build a normalized (flat) topology with parent indexes required for topology create/update
+		normalizedTopology = {
+			name: name,
+			description: description,
+			nodes: []
+		};
+		flattenTopology(topologyTree, true);
+		addNodeIndexes();
+		return normalizedTopology;
+	};
+
+	this.getTopologyTree = function(topology) {
+		let nodes = angular.copy(topology.nodes);
+		let roots = [], // topology items without parents (primary or secondary)
+			all = {};
+
+		nodes.forEach(function(node, index) {
+			all[index] = node;
+		});
```

Comment:
It looks like what you're trying to do is
```javascript
const all = new Map(nodes.map((node,i)=>[i,node]));
// or if webpack won't let you do anonymous "fat-arrow" functions:
const all = new Map(nodes.map(function(node, i){return [i,node];}));
```

That'll allow you to just iterate directly instead of using `Object.keys(all).forEach(function (guid) {` like:
```javascript
for (const [k, v] of all) {
	//...
}
```

Reply:
tried. failed. maybe not as elegant but again pretty simple to read and works.

Reply:
Yeah I actually think our ancient version of webpack won't let you do `for...of` loops. Maybe not even Maps.